### PR TITLE
driver: hwinfo: npcx: Add hwinfo driver support in npcx family.

### PR DESCRIPTION
**Pull request by MulinChao**
Add hwinfo driver support in npcx family.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

### CODE REVIEW

File: drivers/hwinfo/hwinfo_npcx.c [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+ */
+static uint8_t npcx_hwinfo_chip_id(void)
+{
+	return inst_mswc->SRID_CR;
+}
+
+/**
+ * @brief Receive 8-bit value that identifies a specific device of a group.
+ *
+ * @retval value thatidentifies a specific device of a group.
```
Comment (WealianLiao) [inline drivers/hwinfo/hwinfo_npcx.c:40]: that identifies

File: drivers/hwinfo/hwinfo_npcx.c [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+{
+	return inst_mswc->DEVICE_ID_CR;
+}
+
+/**
+ * @brief Receive 8-bit value that identifies the device revision.
+ *
+ * @retval value that identifies the device revision.
+ */
+static uint8_t npcx_hwinfo_revison(void)
```
Comment (WealianLiao) [inline drivers/hwinfo/hwinfo_npcx.c:52]: revision

File: drivers/hwinfo/hwinfo_npcx.c [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+ssize_t z_impl_hwinfo_get_device_id(uint8_t *buffer, size_t length)
+{
+	uint8_t chip_info[4];
+
+	chip_info[0] = npcx_hwinfo_family_id();
+	chip_info[1] = npcx_hwinfo_chip_id();
+	chip_info[2] = npcx_hwinfo_device_id();
+	chip_info[3] = npcx_hwinfo_revison();
```
Comment (WealianLiao) [inline drivers/hwinfo/hwinfo_npcx.c:64]: revision

File: drivers/hwinfo/hwinfo_npcx.c (hunk 1 above)
Comment (MulinChao) [reply drivers/hwinfo/hwinfo_npcx.c]: Done.

File: drivers/hwinfo/hwinfo_npcx.c (hunk 3 above)
Comment (MulinChao) [reply drivers/hwinfo/hwinfo_npcx.c:64]: Done.

File: drivers/hwinfo/hwinfo_npcx.c (hunk 2 above)
Comment (MulinChao) [reply drivers/hwinfo/hwinfo_npcx.c:52]: Done.

**Review by sjg20 — APPROVED** [review]

**Comment by alexanderwachter** [comment]
@MulinChao Thanks for this addition. Do you have a datasheet of this SoC. Do those registers provide unique values? From the naming, it sounds like they identify the chip and are the same for each chip.
The device ID should be a unique ID.

**Review by jackrosenthal — APPROVED** [review]

**Comment by MulinChao** [comment]
> @MulinChao Thanks for this addition. Do you have a datasheet of this SoC. Do those registers provide unique values? From the naming, it sounds like they identify the chip and are the same for each chip.
> The device ID should be a unique ID.
Ping our PM, @Michal-Sch, for further actions.
Hi @alexanderwachter, thanks for your comment. Regarding the datasheet, it is provided under NDA. Please refer to the discussion in [PR27578](https://github.com/zephyrproject-rtos/zephyr/pull/27578#issuecomment-687732361) 

Related to the unique values of these registers, yes, they are unique values that rely on the device's family, group, revision, and so on. They identify the chip's family, group, and revision.

**Comment by alexanderwachter** [comment]
So as I understand, the numbers are unique for each series.
The device id is meant to be unique for each piece. So if you take two pieces of the same type, they have to be different.
For me, as the maintainer of hwinfo, it is hard to give a +1 without having a look at the datasheet.

**Comment by MulinChao** [comment]
> So as I understand, the numbers are unique for each series.
> The device id is meant to be unique for each piece. So if you take two pieces of the same type, they have to be different.
> For me, as the maintainer of hwinfo, it is hard to give a +1 without having a look at the datasheet.

Understood. Sorry for the inconvenience.

**Comment by alexanderwachter** [comment]
Does your MCU support unique IDs?
If not, we should consider closing this PR

**Review by alexanderwachter — CHANGES_REQUESTED** [review]
This driver does not implement a unique id, but an identifier for a series.

**Comment by MulinChao** [comment]
> Does your MCU support unique IDs?
> If not, we should consider closing this PR

If I understand correctly, the unique ID means each different chip in the same series should provide a unique number. If so, I misunderstood 'hwinfo' driver and npcx MCU doesn't support this feature.

**Comment by alexanderwachter** [comment]
As a background:
We are using this API for USB to provide a unique ID. I think updatehub uses it to identify their nodes.

**Comment by MulinChao** [comment]
> As a background:
> We are using this API for USB to provide a unique ID. I think updatehub uses it to identify their nodes.

Got it. Thanks for sharing the background usage. I will close this PR.

**Comment by MulinChao** [comment]
Hi All:
Since the interface of 'hwinfo' is for USB application. I will move it into the other repo and close it.

**Comment by MulinChao** [comment]
Hi All:
Since the interface of 'hwinfo' is for USB application. I will move it into the other repo and close it.

**Pull request closed without merging**